Fix: Add version check for timm to support mobilenetv5 models (fixes #39208)

VIGNESH15103 opened the pull request:
### What does this PR do?

Fixes issue #39208: Adds a version check for `timm` to ensure support for models like `mobilenetv5_300m_enc`, which require `timm >= 0.9.10`.

### Motivation

The issue occurred because the `mobilenetv5_300m_enc` model name was not recognized in older versions of `timm`. This PR adds a conditional check to raise an informative `ImportError` if an unsupported version is detected.

### Changes

- Added a version check using `packaging.version` inside `TimmWrapperImageProcessor.__init__`.
- Raises an error if a `mobilenetv5` model is used with `timm < 0.9.10`.

### Issue Link

Closes #39208

### Reviewer Suggestion

@amyeroberts @qubvel (vision models)

VIGNESH15103 committed:
Fix: Add version check for timm to support mobilenetv5 models (fixes #39208)

VIGNESH15103 committed:
style: fix ruff import order in image_processing_timm_wrapper.py

VIGNESH15103 committed:
Merge branch 'main' into fix-unknown-mobilenetv5

qubvel commented on src/transformers/models/timm_wrapper/image_processing_timm_wrapper.py:
```diff
@@ -59,6 +60,14 @@ def __init__(
         requires_backends(self, "timm")
         super().__init__(architecture=architecture)
 
+        # mobilenetv5 models require timm >= 0.9.10
+        if architecture is not None and architecture.startswith("mobilenetv5"):
+            if version.parse(timm.__version__) < version.parse("0.9.10"):
+                raise ImportError(
+                    f"`mobilenetv5` models require `timm >= 0.9.10`, but found {timm.__version__}. "
+                    "Please upgrade `timm` to the latest version."
+                )
+
```
The version for mibilenetv5 should be `timm >= 1.0.16`

VIGNESH15103 replied:
Thanks @qubvel! Just to confirm — would it be appropriate to add the version check inside TimmWrapperModel.__init__ and TimmWrapperForImageClassification.__init__ right before timm.create_model(...) based on config.architecture? I’ll update accordingly.

qubvel reviewed (commented):
Hey @VIGNESH15103, thanks a lot for the fix, I would suggest we add a check for config loading, not for the image processor

VIGNESH15103 committed:
Raise ImportError when unsupported timm model is used

VIGNESH15103 committed:
Apply ruff formatting to modeling_timm_wrapper.py

VIGNESH15103 committed:
Merge branch 'main' into fix-unknown-mobilenetv5

VIGNESH15103 commented:
Following the feedback:

- I removed the version check from image_processing_timm_wrapper.py.
- Instead, I now catch RuntimeError during timm.create_model(...) inside both TimmWrapperModel and TimmWrapperForImageClassification, as suggested.
- Clear ImportError is raised when the architecture is unsupported, prompting users to upgrade timm.
- Applied ruff formatting to resolve CI failures.

**Kindly requesting a review and workflow approval when convenient. Thanks a lot!**

VIGNESH15103 committed:
Merge branch 'main' into fix-unknown-mobilenetv5

qubvel commented on src/transformers/models/timm_wrapper/modeling_timm_wrapper.py:
```diff
@@ -117,7 +117,15 @@ def __init__(self, config: TimmWrapperConfig):
+            self.timm_model = timm.create_model(
+                config.architecture, pretrained=False, num_classes=0, **extra_init_kwargs
+            )
+        except RuntimeError as e:
+            raise ImportError(
+                f"You are trying to instantiate `{config.architecture}`, but it's not supported in timm={timm.__version__}. "
+                "Please try updating to the latest timm version with `pip install -U timm`."
+            ) from e
```
We should catch an exact exception that the model is unsupported, not a general one

qubvel commented on src/transformers/models/timm_wrapper/image_processing_timm_wrapper.py:
```diff
@@ -59,6 +60,14 @@ def __init__(
         requires_backends(self, "timm")
         super().__init__(architecture=architecture)
 
+        # mobilenetv5 models require timm >= 0.9.10
+        if architecture is not None and architecture.startswith("mobilenetv5"):
+            if version.parse(timm.__version__) < version.parse("0.9.10"):
+                raise ImportError(
+                    f"`mobilenetv5` models require `timm >= 0.9.10`, but found {timm.__version__}. "
+                    "Please upgrade `timm` to the latest version."
```
do we still need this in image processor?

qubvel commented on src/transformers/models/timm_wrapper/modeling_timm_wrapper.py:
```diff
@@ -233,9 +241,16 @@ def __init__(self, config: TimmWrapperConfig):
-        self.timm_model = timm.create_model(
-            config.architecture, pretrained=False, num_classes=config.num_labels, **extra_init_kwargs
-        )
+        try:
+            self.timm_model = timm.create_model(
+                config.architecture, pretrained=False, num_classes=config.num_labels, **extra_init_kwargs
+            )
+        except RuntimeError as e:
+            raise ImportError(
+                f"You are trying to instantiate `{config.architecture}`, but it's not supported in timm=={timm.__version__}. "
+                "Please try updating to the latest timm version with `pip install -U timm`."
+            ) from e
+
```
maybe wrap this in a function

qubvel reviewed (commented):
Thanks for iterating, a few more comments 🤗

VIGNESH15103 committed:
Refactor: Wrap model creation in helper, improve exception handling, remove redundant checks

VIGNESH15103 committed:
Merge branch 'main' into fix-unknown-mobilenetv5

VIGNESH15103 commented:
Hi! I've addressed all your feedback:

-  Wrapped model creation in a helper method as suggested.
-  Added specific exception handling.
-  Removed the redundant version check in the image processor.
-  Ran `ruff` formatting and all CI checks pass now.

Could you please take another look and let me know if there's anything else to improve? If everything looks good, it would be awesome if you could approve and merge it 😊

qubvel commented on src/transformers/models/timm_wrapper/image_processing_timm_wrapper.py:
```diff
@@ -59,6 +59,8 @@ def __init__(
         requires_backends(self, "timm")
         super().__init__(architecture=architecture)
 
+        # mobilenetv5 models require timm >= 0.9.10
+
```
no needed

qubvel commented on src/transformers/models/timm_wrapper/modeling_timm_wrapper.py:
```diff
@@ -117,7 +132,14 @@ def __init__(self, config: TimmWrapperConfig):
         # using num_classes=0 to avoid creating classification head
         extra_init_kwargs = config.model_args or {}
-        self.timm_model = timm.create_model(config.architecture, pretrained=False, num_classes=0, **extra_init_kwargs)
+        try:
+            self.timm_model = safe_create_model(config.architecture, 0, **extra_init_kwargs)
+
+        except RuntimeError as e:
+            raise ImportError(
+                f"You are trying to instantiate `{config.architecture}`, but it's not supported in timm={timm.__version__}. "
+                "Please try updating to the latest timm version with `pip install -U timm`."
+            ) from e
```
we should just call for `safe_create_model` with named args, no need for try catch anymore

qubvel commented on src/transformers/models/timm_wrapper/modeling_timm_wrapper.py:
```diff
@@ -56,6 +56,21 @@ class TimmWrapperModelOutput(ModelOutput):
     attentions: Optional[tuple[torch.FloatTensor, ...]] = None
 
 
+def safe_create_model(architecture, num_classes, **kwargs):
+    try:
+        return timm.create_model(
+            architecture,
+            pretrained=False,
+            num_classes=num_classes,
+            **kwargs,
+        )
+    except RuntimeError as e:
+        raise ImportError(
+            f"You are trying to instantiate `{architecture}`, but it's not supported in timm=={timm.__version__}. "
+            "Please try updating to the latest timm version with `pip install -U timm`."
+        ) from e
```
we should check fora  specific message in the exception, not just catch `RuntimeError`

qubvel reviewed (commented):
Hey! I guess you are using some AI model to create PR, please review changes before submitting them 🤗
